File: tests/dataclasses_json_round_trip.rs
#[path = "support/python_env.rs"]
mod python_env;

use jsoncompat_codegen::generate_dataclass_models;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};
use std::error::Error;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::time::{SystemTime, UNIX_EPOCH};

const FIXTURE_ROOT: &str = "tests/fixtures";
const DATACLASS_SNAPSHOT_ROOT: &str = "tests/fixtures/dataclasses";
const SAMPLE_CACHE: &str = "pybindings/bench_fixture_samples.json";
const DRIVER: &str = include_str!("support/dataclasses_json_round_trip.py");

#[derive(Serialize)]
struct Batch {
    cases: Vec<GeneratedCase>,
}

#[derive(Serialize)]
struct GeneratedCase {
    case_id: String,
    module_path: PathBuf,
    source_path: PathBuf,
    schema_index: Option<usize>,
    expected_schema_digest: Option<String>,
    candidates: Vec<Value>,
    runtime_unsupported: Option<String>,
    unsatisfiable: Option<String>,
}

#[derive(Deserialize)]
struct DriverSummary {
    generated_cases: usize,
    candidates: usize,
    checked_round_trips: usize,
    trusted_round_trips: usize,
    runtime_unsupported: usize,
    unsatisfiable: usize,
}

#[derive(Deserialize)]
struct CachedSample {
    schema_digest: String,
    value: Value,
}

#[derive(Deserialize)]
struct UnsatisfiableFixture {
    reason: String,
    schema_digest: String,
}

struct FixtureCase {
    case_id: String,
    snapshot_base: PathBuf,
    source_path: PathBuf,
    schema_index: Option<usize>,
    schema: Value,
    candidates: Vec<Value>,
}

struct TempDir(PathBuf);

impl TempDir {
    fn create() -> Result<Self, Box<dyn Error>> {
        let unique = SystemTime::now().duration_since(UNIX_EPOCH)?.as_nanos();
        let path = std::env::temp_dir().join(format!(
            "jsoncompat-dataclasses-json-round-trip-{}-{unique}",
            std::process::id(),
        ));
        fs::create_dir_all(&path)?;
        Ok(Self(path))
    }
}

impl Drop for TempDir {
    fn drop(&mut self) {
        let _ = fs::remove_dir_all(&self.0);
    }
}

#[test]
fn every_known_json_schema_fixture_round_trips_through_generated_dataclasses()
-> Result<(), Box<dyn Error>> {
    let repo_root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let fixture_cases = collect_fixture_cases(repo_root)?;
    assert!(
        !fixture_cases.is_empty(),
        "fixture corpus must not be empty"
    );

    let temp_dir = TempDir::create()?;
    let mut generated_cases = Vec::new();
    let mut unsupported_count = 0;
    let mut runtime_unsupported = read_runtime_unsupported(repo_root)?;
    let mut unsatisfiable = read_unsatisfiable(repo_root)?;
    let mut sample_cache = read_sample_cache(repo_root)?;
    let mut case_ids = BTreeSet::new();

    for (index, fixture_case) in fixture_cases.iter().enumerate() {
        assert!(
            case_ids.insert(fixture_case.case_id.clone()),
            "duplicate fixture case id {}",
            fixture_case.case_id,
        );

        match generate_dataclass_models(&fixture_case.schema) {
            Ok(source) => {
                assert_generated_snapshot(repo_root, fixture_case)?;
                let runtime_error = runtime_unsupported.remove(&fixture_case.case_id);
                let unsatisfiable_fixture = unsatisfiable.remove(&fixture_case.case_id);
                assert!(
                    runtime_error.is_none() || unsatisfiable_fixture.is_none(),
                    "fixture {} cannot be both runtime-unsupported and unsatisfiable",
                    fixture_case.case_id,
                );
                let mut candidates = fixture_case.candidates.clone();
                let mut expected_schema_digest = None;
                if runtime_error.is_none() && unsatisfiable_fixture.is_none() {
                    let cached = sample_cache.remove(&fixture_case.case_id).ok_or_else(|| {
                        format!(
                            "runtime-supported satisfiable fixture {} has no checked-in sample in {SAMPLE_CACHE}",
                            fixture_case.case_id,
                        )
                    })?;
                    expected_schema_digest = Some(cached.schema_digest);
                    if !candidates.contains(&cached.value) {
                        candidates.push(cached.value);
                    }
                    assert!(
                        !candidates.is_empty(),
                        "runtime-supported satisfiable fixture {} has no round-trip candidate",
                        fixture_case.case_id,
                    );
                } else {
                    assert!(
                        !sample_cache.contains_key(&fixture_case.case_id),
                        "fixture {} is classified as unsupported but still has a checked-in sample",
                        fixture_case.case_id,
                    );
                }
                if let Some(classification) = &unsatisfiable_fixture {
                    assert!(
                        candidates.is_empty(),
                        "unsatisfiable fixture {} declares a valid candidate",
                        fixture_case.case_id,
                    );
                    assert!(
                        !classification.reason.trim().is_empty(),
                        "unsatisfiable fixture {} has an empty reason",
                        fixture_case.case_id,
                    );
                    expected_schema_digest = Some(classification.schema_digest.clone());
                }
                let module_path = temp_dir.0.join(format!("fixture_{index:04}.py"));
                fs::write(&module_path, source)?;
                generated_cases.push(GeneratedCase {
                    case_id: fixture_case.case_id.clone(),
                    module_path,
                    source_path: fixture_case.source_path.clone(),
                    schema_index: fixture_case.schema_index,
                    expected_schema_digest,
                    candidates,
                    runtime_unsupported: runtime_error,
                    unsatisfiable: unsatisfiable_fixture.map(|entry| entry.reason),
                });
            }
            Err(error) => {
                assert!(
                    !runtime_unsupported.contains_key(&fixture_case.case_id),
                    "fixture {} is classified as both codegen-unsupported and runtime-unsupported",
                    fixture_case.case_id,
                );
                assert!(
                    !unsatisfiable.contains_key(&fixture_case.case_id),
                    "fixture {} is classified as both codegen-unsupported and unsatisfiable",
                    fixture_case.case_id,
                );
                assert!(
                    !sample_cache.contains_key(&fixture_case.case_id),
                    "codegen-unsupported fixture {} still has a checked-in runtime sample",
                    fixture_case.case_id,
                );
                unsupported_count += 1;
                assert_codegen_error_snapshot(repo_root, fixture_case, &error.to_string())?;
            }
        }
    }

    assert_eq!(
        generated_cases.len() + unsupported_count,
        fixture_cases.len(),
        "every fixture schema must be generated or explicitly classified as unsupported",
    );
    assert!(
        runtime_unsupported.is_empty(),
        "runtime-unsupported classifications do not identify generated fixture cases: {:?}",
        runtime_unsupported.keys().collect::<Vec<_>>(),
    );
    assert!(
        unsatisfiable.is_empty(),
        "unsatisfiable classifications do not identify generated fixture cases: {:?}",
        unsatisfiable.keys().collect::<Vec<_>>(),
    );
    assert!(
        sample_cache.is_empty(),
        "checked-in samples do not identify runtime-supported satisfiable generated fixtures: {:?}",
        sample_cache.keys().collect::<Vec<_>>(),
    );

    let candidate_count = generated_cases
        .iter()
        .filter(|case| case.runtime_unsupported.is_none() && case.unsatisfiable.is_none())
        .map(|case| case.candidates.len())
        .sum::<usize>();
    let runtime_unsupported_count = generated_cases
        .iter()
        .filter(|case| case.runtime_unsupported.is_some())
        .count();
    let unsatisfiable_count = generated_cases
        .iter()
        .filter(|case| case.unsatisfiable.is_some())
        .count();
    let expected_generated_count = generated_cases.len();
    let mut child = python_env::python_command();
    child
        .args(["-B", "-c", DRIVER])
        .env_remove("PYTHONSAFEPATH")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    let mut child = child.spawn()?;
    serde_json::to_writer(
        child.stdin.as_mut().expect("piped Python stdin"),
        &Batch {
            cases: generated_cases,
        },
    )?;
    child.stdin.take().expect("piped Python stdin").flush()?;

    let output = child.wait_with_output()?;
    assert!(
        output.status.success(),
        "batched generated-dataclass JSON round-trip failed\n\nstdout:\n{}\n\nstderr:\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr),
    );

    let summary: DriverSummary = serde_json::from_slice(&output.stdout).map_err(|error| {
        format!(
            "round-trip driver returned an invalid summary: {error}\nstdout:\n{}\nstderr:\n{}",
            String::from_utf8_lossy(&output.stdout),
            String::from_utf8_lossy(&output.stderr),
        )
    })?;
    assert_eq!(summary.generated_cases, expected_generated_count);
    assert_eq!(summary.candidates, candidate_count);
    assert_eq!(summary.checked_round_trips, candidate_count);
    assert_eq!(summary.trusted_round_trips, candidate_count);
    assert_eq!(summary.runtime_unsupported, runtime_unsupported_count);
    assert_eq!(summary.unsatisfiable, unsatisfiable_count);

    eprintln!(
        "generated {expected_generated_count}/{} fixture schemas; codegen-unsupported: {unsupported_count}; runtime-unsupported: {runtime_unsupported_count}; unsatisfiable: {unsatisfiable_count}; checked and trusted JSON round-trips: {candidate_count} each",
        fixture_cases.len(),
    );
    Ok(())
}

fn read_unsatisfiable(
    repo_root: &Path,
) -> Result<BTreeMap<String, UnsatisfiableFixture>, Box<dyn Error>> {
    Ok(serde_json::from_slice(&fs::read(
        repo_root
            .join(DATACLASS_SNAPSHOT_ROOT)
            .join("unsatisfiable.json"),
    )?)?)
}

fn read_sample_cache(repo_root: &Path) -> Result<BTreeMap<String, CachedSample>, Box<dyn Error>> {
    Ok(serde_json::from_slice(&fs::read(
        repo_root.join(SAMPLE_CACHE),
    )?)?)
}

fn read_runtime_unsupported(repo_root: &Path) -> Result<BTreeMap<String, String>, Box<dyn Error>> {
    let path = repo_root
        .join(DATACLASS_SNAPSHOT_ROOT)
        .join("runtime_unsupported.json");
    read_nonempty_string_manifest(&path, "runtime-unsupported")
}

fn read_nonempty_string_manifest(
    path: &Path,
    name: &str,
) -> Result<BTreeMap<String, String>, Box<dyn Error>> {
    let classifications: BTreeMap<String, String> = serde_json::from_slice(&fs::read(path)?)?;
    assert!(
        classifications
            .values()
            .all(|reason| !reason.trim().is_empty()),
        "{name} classifications in {} must include a non-empty reason",
        path.display(),
    );
    Ok(classifications)
}

fn collect_fixture_cases(repo_root: &Path) -> Result<Vec<FixtureCase>, Box<dyn Error>> {
    let fixture_root = repo_root.join(FIXTURE_ROOT);
    let mut cases = collect_backcompat_cases(&fixture_root.join("backcompat"))?;
    cases.extend(collect_fuzz_cases(&fixture_root.join("fuzz"))?);
    cases.sort_by(|left, right| left.case_id.cmp(&right.case_id));
    Ok(cases)
}

fn collect_backcompat_cases(root: &Path) -> Result<Vec<FixtureCase>, Box<dyn Error>> {
    let mut cases = Vec::new();
    for fixture_dir in sorted_dirs(root)? {
        let case_name = utf8_file_name(&fixture_dir)?;
        let examples = read_optional_json(&fixture_dir.join("examples.json"))?;
        for side in ["old", "new"] {
            let relative_base = Path::new("backcompat").join(&case_name).join(side);
            let source_path = fixture_dir.join(format!("{side}.json"));
            cases.push(FixtureCase {
                case_id: relative_base.to_string_lossy().replace('\\', "/"),
                snapshot_base: relative_base,
                schema: read_json(&source_path)?,
                source_path,
                schema_index: None,
                candidates: backcompat_candidates(examples.as_ref(), side)?,
            });
        }
    }
    Ok(cases)
}

fn backcompat_candidates(
    examples: Option<&Value>,
    side: &str,
) -> Result<Vec<Value>, Box<dyn Error>> {
    let Some(examples) = examples else {
        return Ok(Vec::new());
    };
    let object = examples
        .as_object()
        .ok_or("backcompat examples must be a JSON object")?;
    let mut candidates = Vec::new();
    for key in ["both".to_owned(), format!("{side}_only")] {
        let Some(values) = object.get(&key) else {
            continue;
        };
        let values = values
            .as_array()
            .ok_or_else(|| format!("backcompat examples field {key:?} must be an array"))?;
        candidates.extend(values.iter().cloned());
    }
    Ok(candidates)
}

fn collect_fuzz_cases(root: &Path) -> Result<Vec<FixtureCase>, Box<dyn Error>> {
    let mut cases = Vec::new();
    for fixture_path in sorted_json_files(root)? {
        let relative_path = fixture_path.strip_prefix(root)?;
        let snapshot_dir = Path::new("fuzz").join(relative_path.with_extension(""));
        let root_value = read_json(&fixture_path)?;
        for (index, (schema, candidates)) in
            embedded_fuzz_schemas(&root_value).into_iter().enumerate()
        {
            let snapshot_base = snapshot_dir.join(format!("{index:03}"));
            cases.push(FixtureCase {
                case_id: snapshot_base.to_string_lossy().replace('\\', "/"),
                snapshot_base,
                source_path: fixture_path.clone(),
                schema_index: Some(index),
                schema,
                candidates,
            });
        }
    }
    Ok(cases)
}

fn embedded_fuzz_schemas(root: &Value) -> Vec<(Value, Vec<Value>)> {
    let Value::Array(items) = root else {
        return vec![(root.clone(), Vec::new())];
    };
    items
        .iter()
        .filter_map(|item| {
            let schema = item.get("schema")?.clone();
            let candidates = item
                .get("tests")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
                .filter(|test| test.get("valid").and_then(Value::as_bool) == Some(true))
                .filter_map(|test| test.get("data").cloned())
                .collect();
            Some((schema, candidates))
        })
        .collect()
}

fn assert_generated_snapshot(
    repo_root: &Path,
    fixture_case: &FixtureCase,
) -> Result<(), Box<dyn Error>> {
    let python_path = snapshot_path(repo_root, &fixture_case.snapshot_base, "py");
    let error_path = snapshot_path(repo_root, &fixture_case.snapshot_base, "error.txt");
    if !python_path.is_file() || error_path.exists() {
        return Err(format!(
            "generated fixture {} must have exactly one Python snapshot at {}; run `just regen-dataclasses-fixtures`",
            fixture_case.case_id,
            python_path.display(),
        )
        .into());
    }
    Ok(())
}

fn assert_codegen_error_snapshot(
    repo_root: &Path,
    fixture_case: &FixtureCase,
    error: &str,
) -> Result<(), Box<dyn Error>> {
    let error_path = snapshot_path(repo_root, &fixture_case.snapshot_base, "error.txt");
    let python_path = snapshot_path(repo_root, &fixture_case.snapshot_base, "py");
    if python_path.exists() {
        return Err(format!(
            "unsupported fixture {} still has a generated Python snapshot {}; run `just regen-dataclasses-fixtures`",
            fixture_case.case_id,
            python_path.display(),
        )
        .into());
    }
    let expected = fs::read_to_string(&error_path).map_err(|read_error| {
        format!(
            "fixture {} failed code generation without an explicit classification at {}: {read_error}",
            fixture_case.case_id,
            error_path.display(),
        )
    })?;
    let actual = format!("{error}\n");
    if normalized_newlines(&expected) != normalized_newlines(&actual) {
        return Err(format!(
            "codegen error classification is stale for {} at {}\n\nexpected:\n{}\nactual:\n{}",
            fixture_case.case_id,
            error_path.display(),
            expected,
            actual,
        )
        .into());
    }
    Ok(())
}

fn snapshot_path(repo_root: &Path, relative_base: &Path, extension: &str) -> PathBuf {
    repo_root
        .join(DATACLASS_SNAPSHOT_ROOT)
        .join(relative_base)
        .with_extension(extension)
}

fn normalized_newlines(contents: &str) -> String {
    contents.replace("\r\n", "\n")
}

fn read_json(path: &Path) -> Result<Value, Box<dyn Error>> {
    Ok(serde_json::from_slice(&fs::read(path)?)?)
}

fn read_optional_json(path: &Path) -> Result<Option<Value>, Box<dyn Error>> {
    if path.is_file() {
        Ok(Some(read_json(path)?))
    } else {
        Ok(None)
    }
}

fn sorted_dirs(root: &Path) -> Result<Vec<PathBuf>, Box<dyn Error>> {
    let mut paths = fs::read_dir(root)?
        .map(|entry| entry.map(|entry| entry.path()))
        .collect::<Result<Vec<_>, _>>()?;
    paths.retain(|path| path.is_dir());
    paths.sort();
    Ok(paths)
}

fn sorted_json_files(root: &Path) -> Result<Vec<PathBuf>, Box<dyn Error>> {
    let mut files = Vec::new();
    let mut pending = vec![root.to_path_buf()];
    while let Some(path) = pending.pop() {
        if path.is_dir() {
            pending.extend(
                fs::read_dir(path)?
                    .map(|entry| entry.map(|entry| entry.path()))
                    .collect::<Result<Vec<_>, _>>()?,
            );
        } else if path.extension().and_then(|extension| extension.to_str()) == Some("json") {
            files.push(path);
        }
    }
    files.sort();
    Ok(files)
}

fn utf8_file_name(path: &Path) -> Result<String, Box<dyn Error>> {
    path.file_name()
        .and_then(|name| name.to_str())
        .map(str::to_owned)
        .ok_or_else(|| format!("fixture path is not valid UTF-8: {}", path.display()).into())
}

File: tests/support/dataclasses_json_round_trip.py
"""Batched JSON round-trip driver for every generated dataclass fixture."""

from __future__ import annotations

import hashlib
import importlib.util
import json
import sys
from dataclasses import is_dataclass
from pathlib import Path
from types import ModuleType
from typing import Any, TypedDict, cast


type JsonScalar = None | bool | int | float | str
type JsonValue = JsonScalar | list["JsonValue"] | dict[str, "JsonValue"]


class FixtureCase(TypedDict):
    case_id: str
    module_path: str
    source_path: str
    schema_index: int | None
    expected_schema_digest: str | None
    candidates: list[JsonValue]
    runtime_unsupported: str | None
    unsatisfiable: str | None


class FixtureBatch(TypedDict):
    cases: list[FixtureCase]


def load_module(case_id: str, module_path: Path, index: int) -> ModuleType:
    module_name = f"_jsoncompat_fixture_round_trip_{index:04}"
    spec = importlib.util.spec_from_file_location(module_name, module_path)
    if spec is None or spec.loader is None:
        raise AssertionError(f"{case_id}: could not create an import spec")
    module = importlib.util.module_from_spec(spec)
    sys.modules[module_name] = module
    try:
        spec.loader.exec_module(module)
    except BaseException:
        sys.modules.pop(module_name, None)
        raise
    return module


def json_equivalent(left: JsonValue, right: JsonValue) -> bool:
    if isinstance(left, dict) and isinstance(right, dict):
        return left.keys() == right.keys() and all(
            json_equivalent(left[key], right[key]) for key in left
        )
    if isinstance(left, list) and isinstance(right, list):
        return len(left) == len(right) and all(
            json_equivalent(left_item, right_item)
            for left_item, right_item in zip(left, right, strict=True)
        )
    if isinstance(left, bool) or isinstance(right, bool):
        return isinstance(left, bool) and isinstance(right, bool) and left == right
    if isinstance(left, (int, float)) and isinstance(right, (int, float)):
        return left == right
    return type(left) is type(right) and left == right


def canonical_json(value: JsonValue) -> str:
    return json.dumps(
        value,
        allow_nan=False,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    )


def fixture_schema(case: FixtureCase) -> JsonValue:
    root = cast(JsonValue, json.loads(Path(case["source_path"]).read_text()))
    schema_index = case["schema_index"]
    if schema_index is None:
        return root
    if not isinstance(root, list):
        raise AssertionError(f"{case['case_id']}: indexed fixture is not an array")
    item = root[schema_index]
    if not isinstance(item, dict) or "schema" not in item:
        raise AssertionError(f"{case['case_id']}: fixture item has no schema")
    return item["schema"]


def assert_round_trip(
    *,
    case_id: str,
    candidate_index: int,
    candidate: JsonValue,
    model_type: Any,
    trusted: bool,
) -> None:
    wire = canonical_json(candidate)
    try:
        model = model_type.deserialize(wire, skip_validation=trusted)
        if not isinstance(model, model_type) or not is_dataclass(model):
            raise AssertionError(
                f"{case_id} candidate #{candidate_index}: deserialize returned "
                f"{type(model).__name__}, expected a {model_type.__name__} dataclass"
            )
        emitted_wire = cast(Any, model).serialize(skip_validation=trusted)
        emitted = cast(JsonValue, json.loads(emitted_wire))
    except BaseException as error:
        mode = "trusted" if trusted else "checked"
        raise AssertionError(
            f"{case_id} candidate #{candidate_index}: {mode} JSON round-trip "
            f"failed for {wire}: {type(error).__name__}: {error}"
        ) from error
    if not json_equivalent(candidate, emitted):
        mode = "trusted" if trusted else "checked"
        raise AssertionError(
            f"{case_id} candidate #{candidate_index}: {mode} JSON round-trip "
            f"changed the value: {wire} -> {canonical_json(emitted)}"
        )


def assert_runtime_unsupported(
    *,
    case_id: str,
    candidates: list[JsonValue],
    model_type: Any,
    expected_error: str,
) -> None:
    if not candidates:
        raise AssertionError(
            f"{case_id}: runtime-unsupported classification has no fixture value "
            "with which to verify the expected error"
        )
    wire = canonical_json(candidates[0])
    try:
        model_type.deserialize(wire)
    except BaseException as error:
        actual = f"{type(error).__name__}: {error}"
        if expected_error != actual:
            raise AssertionError(
                f"{case_id}: runtime-unsupported classification is stale; "
                f"expected {expected_error!r}, got {actual!r}"
            ) from error
    else:
        raise AssertionError(
            f"{case_id}: runtime-unsupported fixture now deserializes; remove its "
            "classification and exercise its complete fixture corpus"
        )


def main() -> None:
    batch = cast(FixtureBatch, json.load(sys.stdin))
    cases = batch["cases"]
    candidate_count = 0
    runtime_unsupported_count = 0
    unsatisfiable_count = 0
    failures: list[str] = []
    for case_index, case in enumerate(cases):
        case_id = case["case_id"]
        expected_runtime_error = case["runtime_unsupported"]
        unsatisfiable_reason = case["unsatisfiable"]
        expected_schema_digest = case["expected_schema_digest"]
        if expected_schema_digest is not None:
            actual_schema_digest = hashlib.sha256(
                canonical_json(fixture_schema(case)).encode()
            ).hexdigest()
            if actual_schema_digest != expected_schema_digest:
                failures.append(
                    f"{case_id}: checked-in classification/sample schema digest is "
                    f"stale; expected {actual_schema_digest}, got "
                    f"{expected_schema_digest}"
                )
                continue
        try:
            module = load_module(case_id, Path(case["module_path"]), case_index)
            model_type = module.JSONCOMPAT_MODEL
        except BaseException as error:
            actual = f"{type(error).__name__}: {error}"
            if expected_runtime_error == actual:
                runtime_unsupported_count += 1
            elif expected_runtime_error is None:
                failures.append(f"{case_id}: generated module import failed: {actual}")
            else:
                failures.append(
                    f"{case_id}: runtime-unsupported classification is stale; "
                    f"expected {expected_runtime_error!r}, got {actual!r}"
                )
            continue
        if expected_runtime_error is not None:
            runtime_unsupported_count += 1
            try:
                assert_runtime_unsupported(
                    case_id=case_id,
                    candidates=case["candidates"],
                    model_type=model_type,
                    expected_error=expected_runtime_error,
                )
            except AssertionError as error:
                failures.append(str(error))
            continue
        if unsatisfiable_reason is not None:
            if case["candidates"]:
                failures.append(
                    f"{case_id}: unsatisfiable fixture declares valid candidates"
                )
            else:
                unsatisfiable_count += 1
            continue
        for candidate_index, candidate in enumerate(case["candidates"]):
            candidate_count += 1
            try:
                assert_round_trip(
                    case_id=case_id,
                    candidate_index=candidate_index,
                    candidate=candidate,
                    model_type=model_type,
                    trusted=False,
                )
                assert_round_trip(
                    case_id=case_id,
                    candidate_index=candidate_index,
                    candidate=candidate,
                    model_type=model_type,
                    trusted=True,
                )
            except AssertionError as error:
                failures.append(str(error))

    if failures:
        raise AssertionError(
            f"{len(failures)} generated-dataclass fixture round-trip failures:\n"
            + "\n".join(f"- {failure}" for failure in failures)
        )

    print(
        canonical_json(
            {
                "candidates": candidate_count,
                "checked_round_trips": candidate_count,
                "generated_cases": len(cases),
                "runtime_unsupported": runtime_unsupported_count,
                "trusted_round_trips": candidate_count,
                "unsatisfiable": unsatisfiable_count,
            }
        )
    )


if __name__ == "__main__":
    main()
